File: Parallel-lang/lexer.swift
import Foundation

func lex() -> [token] {
	var tokens: [token] = []
	
	var line = 0
	var column = 0
	
	var index = 0
	
	while (true) {
		if (sourceCode.count <= index) {
			break
		}
		let char = sourceCode[sourceCode.index(sourceCode.startIndex, offsetBy: index)]
		
		let startColumn = column
		
		if (char == "\n") {
			line += 1
			column = 0
			
			index += 1
			continue
		}
		
		// space or tab
		else if (char == " " || char == "\t") {
			// do nothing
		}
		
		// words must start with a letter or an underscore
		else if (char.isLetter || char == "_") {
			
			let value = readWhile({ char in
				// words can have numbers after the first character
				return char.isLetter || char == "_" || char.isNumber
			})
			
			tokens.append(token(
				name: "word",
				lineNumber: line,
				start: startColumn,
				end: column,
				value: value
			))
		}
		
		// strings
		else if (char == "\"") {
			// eat the opening "
			column += 1
			index += 1
			
			var value = ""
			
			var escaped = false
			var continueLoop = true
			
			while (continueLoop) {
				if (sourceCode.count <= index) {
					column -= 1
					index -= 1
					continueLoop = false
				} else {
					let char = sourceCode[sourceCode.index(sourceCode.startIndex, offsetBy: index)]
					
					if (escaped) {
						escaped = false
						
						column += 1
						index += 1
						
						if (char == "n") {
							value.append("\n")
							continue
						} else if (char == "\"") {
							continue
						}
					}
					
					if (char == "\\") {
						escaped = true
						column += 1
						index += 1
					} else if (char != "\"") {
						value.append(char)
						if (char == "\n") {
							line += 1
							column = 0
						} else {
							column += 1
							index += 1
						}
					} else {
						column -= 1
						index -= 1
						continueLoop = false
					}
				}
			}
			
			tokens.append(token(
				name: "string",
				lineNumber: line,
				start: startColumn,
				end: column,
				value: value
			))
			
			// eat the closing "
			column += 1
			index += 1
		}
		
		else if (char.isNumber) {
			tokens.append(token(
				name: "number",
				lineNumber: line,
				start: startColumn,
				end: column,
				value: readWhile({ char in
					return char.isNumber || char == "."
				})
			))
		}
		
		else if (char == "(" || char == ")" || char == "{" || char == "}" || char == ":") {
			tokens.append(token(
				name: "separator",
				lineNumber: line,
				start: startColumn,
				end: column,
				value: String(char)
			))
		}
		
		else if (char == "=") {
			tokens.append(token(
				name: "operator",
				lineNumber: line,
				start: startColumn,
				end: column,
				value: String(char)
			))
		}
		
		// for comments
		else if (char == "/" && sourceCode[sourceCode.index(sourceCode.startIndex, offsetBy: index+1)] == "/") {
			let _ = readWhile({ char in
				return char != "\n"
			})
		}
		
		/*
			for multi line comments
		*/
		else if (char == "/" && sourceCode[sourceCode.index(sourceCode.startIndex, offsetBy: index+1)] == "*") {
			let _ = readWhile({ char in
				if (sourceCode.count <= index) {
					compileError("multi line comment never ended")
				}
				let next = sourceCode[sourceCode.index(sourceCode.startIndex, offsetBy: index+1)]
				return char != "*" || next != "/"
			})
			
			// eat the / and *
			column += 2
			index += 2
		}
		
		else {
			compileError("unexpected character: \(char)", line, column, column)
		}
		
		column += 1
		index += 1
	}
	
	return tokens
	
	func readWhile(_ function: (_ char: Character) -> Bool) -> String {
		var past = ""
		while (true) {
			if (sourceCode.count <= index) {
				column -= 1
				index -= 1
				return past
			} else {
				let char = sourceCode[sourceCode.index(sourceCode.startIndex, offsetBy: index)]
				
				if (function(char)) {
					past.append(char)
					if (char == "\n") {
						line += 1
						column = 0
					} else {
						column += 1
					}
					index += 1
				} else {
					column -= 1
					index -= 1
					return past
				}
			}
		}
	}
}

File: Parallel-lang/lexer.c
#include <stdio.h>
#include <string.h>

#include "lexer.h"
#include "globals.h"
#include "error.h"

#define wordStart (character >= 'a' && character <= 'z') || (character >= 'A' && character <= 'Z') || character == '_'
#define wordContinue wordStart || (character >= '0' && character <= '9')

#define numberStart (character >= '0' && character <= '9')
#define numberContinue (character >= '0' && character <= '9')

// periods are considered operators because they perform an operation
#define operator_1char character == '>' || character == '<' || character == '=' || character == '+' || character == '-' || character == '.'
#define operator_2chars character == '=' && MI->currentSource[index+1] == '=' || character == ':' && MI->currentSource[index+1] == ':'

#define separator character == '(' || character == ')' || character == '{' || character == '}'  || character == '[' || character == ']' || character == ':' || character == ';' || character == ','

linkedList_Node *lex(ModuleInformation *MI) {
	linkedList_Node *tokens = 0;
	
	int index = 0;
	int line = 1;
	int column = 0;
	
	if (MI->currentSource == NULL) {
		return NULL;
	}
	
	while (1) {
		char character = MI->currentSource[index];
		
		// if character is a NULL byte, then the string is over
		if (character == 0) {
			break;
		}
		
//		printf("character: %c\n", character);
		
		if (character == '\n') {
			line++;
			column = 0;
			
			index++;
			continue;
		}
		
		// ignore space and tab
		else if (character == ' ' || character == '\t') {}
		
		// start of a word
		else if (wordStart) {
			int start = index;
			int columnStart = column;
			int end = 0;
			
			while (1) {
				char character = MI->currentSource[index];
				
				if (character != 0 && wordContinue) {
					index++;
					column++;
					continue;
				}
				
				end = index;
				
				index--;
				column--;
				break;
			}
			
			Token *data = linkedList_addNode(&tokens, sizeof(Token));
			
			data->type = TokenType_word;
			data->location = (SourceLocation){line, columnStart, columnStart + end - start};
			data->subString.start = MI->currentSource + start;
			data->subString.length = end - start;
		}
		
		else if (operator_2chars) {
			Token *data = linkedList_addNode(&tokens, sizeof(Token));
			
			data->type = TokenType_operator;
			data->location = (SourceLocation){line, column, column + 2};
			data->subString.start = MI->currentSource + index;
			data->subString.length = 2;
			
			index++;
			column++;
		}
		
		else if (operator_1char) {
			Token *data = linkedList_addNode(&tokens, sizeof(Token));
			
			data->type = TokenType_operator;
			data->location = (SourceLocation){line, column, column + 1};
			data->subString.start = MI->currentSource + index;
			data->subString.length = 1;
		}
		
		else if (separator) {
			Token *data = linkedList_addNode(&tokens, sizeof(Token));
			
			data->type = TokenType_separator;
			data->location = (SourceLocation){line, column, column + 1};
			data->subString.start = MI->currentSource + index;
			data->subString.length = 1;
		}
		
		else if (numberStart) {
			int start = index;
			int columnStart = column;
			int end = 0;
			
			while (1) {
				char character = MI->currentSource[index];
				
				if (character != 0 && numberContinue) {
					index++;
					column++;
					continue;
				}
				
				end = index;
				
				index--;
				column--;
				break;
			}
			
			Token *data = linkedList_addNode(&tokens, sizeof(Token));
			
			data->type = TokenType_number;
			data->location = (SourceLocation){line, columnStart, columnStart + end - start};
			data->subString.start = MI->currentSource + start;
			data->subString.length = end - start;
		}
		
		else if (character == '"') {
			// eat the starting quotation mark
			index++;
			column++;
			
			int start = index;
			int columnStart = column;
			int end = 0;
			
			while (1) {
				char character = MI->currentSource[index];
				
				if (character != 0 && character != '"') {
					index++;
					column++;
					if (character == '\n') {
						line++;
						column = 0;
					}
					continue;
				}
				
				end = index;
				
				index--;
				column--;
				
				if (character == '"') {
					// eat the ending quotation mark
					index++;
					column++;
				}
				break;
			}
			
			Token *data = linkedList_addNode(&tokens, sizeof(Token));
			
			data->type = TokenType_string;
			data->location = (SourceLocation){line, columnStart - 1, columnStart + end - start + 1};
			data->subString.start = MI->currentSource + start;
			data->subString.length = end - start;
		}
		
		// comment
		else if (character == '/' && MI->currentSource[index+1] == '/') {
			index++;
			column++;
			
			while (1) {
				char character = MI->currentSource[index];
				
				if (character != 0 && character != '\n') {
					index++;
					column++;
					continue;
				}
				
				index--;
				column--;
				break;
			}
		}
		
		// block comment
		else if (character == '/' && MI->currentSource[index+1] == '*') {
			index++;
			column++;
			
			while (1) {
				char character = MI->currentSource[index];
				
				if (character == 0) {
					index--;
					column--;
					break;
				}
				
				if (character == '\n') {
					line++;
				} else if (character == '*' && MI->currentSource[index+1] == '/') {
					index++;
					column++;
					break;
				}
				
				index++;
				column++;
			}
		}
		
		else {
			printf("unexpected character '%c'\n", character);
			compileError(MI, (SourceLocation){line, column, column + 1});
		}
		
		index++;
		column++;
	}
	
	return tokens;
}
